apps/arweave/src/ar_merkle.erl

```diff
@@ -208,81 +304,96 @@ extract_root(_) ->
 %%% Private functions.
 %%%===================================================================
 
-generate_leaves(Elements) ->
-	lists:foldr(
-		fun({Data, Note}, Nodes) ->
-			Hash = hash([hash(Data), hash(note_to_binary(Note))]),
-			insert(
-				#node {
-					id = Hash,
-					type = leaf,
-					data = Data,
-					note = Note,
-					max = Note
-				},
-				Nodes
-			)
-		end,
-		new(),
-		Elements
-	).
+generate_tree([Element | Elements], Stack, Tree) when is_list(Element) ->
+	{SubRoot, SubTree} = generate_tree(Element),
+	SubTree2 = [mark_rebased(Node, SubRoot) || Node <- SubTree],
+	SubRootN = get(SubRoot, SubTree2),
+	generate_tree(Elements, queue:in(SubRootN, Stack), Tree ++ SubTree2);
+generate_tree([Element | Elements], Stack, Tree) ->
+	Leaf = generate_leaf(Element),
+	generate_tree(Elements, queue:in(Leaf, Stack), [Leaf | Tree]);
```

Comment:
Why is SubTree2 added to the end of the Tree, but Leaf is added to the beginning?

Reply:
The tree structure here is not sensitive to the ordering so it makes a tiny bit more sense to prepend (as it is faster). SubTree is a list too though so it does not matter.